### PR TITLE
[MRG+1] Simplify _reshape_2D.

### PR DESCRIPTION
**Pull request by anntzer**
(I still think iterating over rows vs. over columns depending on input
type is rather silly but at least this implementation makes it clearer.)

(Written while working on #8092, attn @phobson)

### CODE REVIEW

**Comment by phobson** [comment]
👍  Test failure looks unrelated. 

I also notice that `_reshape_2d` doesn't have any of its own tests. I can take a stab at those and push to this PR if you want.

**Comment by anntzer** [comment]
It is implicitly tested by the various input shapes that tests pass to `boxplot` (which includes 2D array, ragged arrays and whatever weirdness).

File: lib/matplotlib/cbook/__init__.py [hunk 1]
```diff
@@ -1983,32 +1983,16 @@ def _reshape_2D(X):
...
-                X = [X.ravel()]
-            else:
-                X = [X[:, i] for i in xrange(ncols)]
-        else:
-            raise ValueError("input `X` must have 2 or fewer dimensions")
-
-    if not hasattr(X[0], '__len__'):
-        X = [X]
+    # Iterate over columns for ndarrays, over rows otherwise.
+    X = X.T if isinstance(X, np.ndarray) else np.asarray(X)
```
Comment (NelleV) [inline lib/matplotlib/cbook/__init__.py]: Why would you behave differently for ndarrays than for any other 2D types? (such as matrices, 2D lists etc.)

Comment (anntzer) [inline lib/matplotlib/cbook/__init__.py]: See https://github.com/matplotlib/matplotlib/issues/8092#issuecomment-280890239.

Comment (NelleV) [inline lib/matplotlib/cbook/__init__.py]: I feel very uncomfortable with having such a behavior in something that could be easily called in functions where we wouldn't want this behaviour. Is it possible to rename this function to something like `_fundamentally_broken_reshape_2D` and have a huge warning in the documentation?

Comment (NelleV) [inline lib/matplotlib/cbook/__init__.py]: I would even go as far as raising a warning when a list is provided and deprecate the behavior we have with list.

Comment (anntzer) [inline lib/matplotlib/cbook/__init__.py]: I agree with you, but I think these should be a separate issue.  The only point of this PR is to deobfuscate the implementation of this function.

I think it is important to keep the ability to pass lists in -- it is natural to plot boxplots with unequal sized datasets, and you can't do that with a 2D array.

**Comment by NelleV** [comment]
@tacaswell can we discuss about this PR tomorrow?

**Pull request merged**